### ISSUE TITLE
Use clang for parser

### ISSUE THREAD

**Issue by ReDucTor** [swig/swig#1159]
SWIG is constantly playing catchup with the new standards of C++, it would be good to look at using clang for parsing C/C++ as it as an excellent tooling API.

One thing that we need to be aware of is that SWIG can work when include files are missing (clang format is also in a similar situation)

The custom parser and constant fighting for new standards is sometimes toughted as why [the original SWIG author hates SWIG](http://code.activestate.com/lists/python-dev/109281/)

**Comment by vadz** [comment]
I thought about replacing SWIG parser with libclang a few years ago, but there were several problems which seemed difficult to overcome and I think only one of them (I needed to parse comments which were not supported by libclang back then, but, I believe, is now) has been solved. Some of the remaining ones:

1. We need to parse SWIG directives and not just C++ code. I have no idea about how could this be integrated with libclang approach.
1. The AST/DOM/whatever created by the current parser is not at all isomorphic to the one used by libclang. Changing all the existing code to use it seems impossible. Maintaining 2 parsers/DOMs in parallel for the old and new code -- even less so.
1. This is a consequence/generalization of the above, but IME SWIG project unfortunately just doesn't have the resources to integrate a change of this magnitude. Even if somebody manages to make some backend(s) to work with libclang instead of the current parser, there is no certainty that this work is going to be merged.

Personally, I think that there are very few chances of this happening with the current codebase. Writing a new project doing the same thing as SWIG (maybe even using the same/compatible typemaps) but using libclang underneath seems like a more realistic -- even if still hardly trivial -- project.
